README
``Panel Discussion.pdf'' and ``Final Copy.docx'' consist of an introductory script, app overview and course material.
``Panel Discussion.pdf'' also begins with a script for inviting experts to participate in the study and ends with a feedback form template.

Instructional App = course material + iOS app

How to Replicate the Panel Discussion
-------------------------------------

Invite experts to participate in the study.
When one meets with the participants, follow the introductory script. Keep the app overview for personal reference.
Course material is distributed to experts for them to complete. However, do not simply hand them the material; illustrate some examples briefly.
Once experts are done, ask them to provide feedback.

How to Utilize the Final Copy
-----------------------------

Instructors follow the introductory script and can keep the app overview for their personal reference.
Course material is distributed to students for them to complete. However, instructors should not simply hand them the material; they should illustrate some examples briefly.
On how to proceed from there is the instructor's discretion.

[Notes]

Jupyter/Ten-Year_Minimum_Errors.ipynb (for computing ten-year minimum monthly payment errors)
LaTeX/deeper_insight.pdf (provides context, overview, terminology and formulae)
LaTeX/deeper_insight.tex (source code)
LaTeX/extra_insight.pdf (provides examples and derivations)
LaTeX/extra_insight.tex (source code)
LaTeX/images/ (images for deeper_insight.pdf)
LaTeX/images/svg/ (source of images for deeper_insight.pdf)
Other/rounding_notes (two notes about rounding quantities in Swift)
Other/checking_calculations.xlsx (for checking app calculations)
Other/checklist_for_checking_calculations.docx (checklist for checking app calculations)
Video/introduction.mov (What is shown, when app launches)

``Video'' directory contains instructions on how to export a new introduction.mov, as well.
Utilizing a separate CHANGELOG for the ``Video'' directory because the video itself is drawn from a project.

CHANGELOG
Nov 1, 2020 · Version 1.7.1: removed mention of sagemath
Oct 20, 2020 · Version 1.7: upgraded Jupyter Notebook file code to Python 3, see Ten-Year_Minimum_Errors.py; fixed issue in which progress would not always show periodically
Aug 2, 2020 · Version 1.6.4: proofread deeper_insight.tex
Mar 24, 2020 · Version 1.6.3: changed "Jupyter Notebook or Jupyter Lab" to "SageMath Notebook"
Nov 30, 2019 · Version 1.6.2: fixed latex images
Sep 25, 2019: unified author name and email of all commits
Sep 24-28, 2019 · Version 1.6.1: extracted some resources from the "course material" repository into a newly separate repository, added extension to rounding_notes, renamed rounding_notes, archived the resources
Aug 27, 2019 · Version 1.6: updated interest rate and default numbers, aligned how savings and change in savings are calculated and displayed
Aug 20, 2019 · Version 1.5: reorganized resources again, corrected spreadsheet
Aug 8, 2019 · Version 1.4: reorganized resources, provided raw files for images, added README files
Jul 29-Aug 5, 2019 · Version 1.3: uploaded Ten-Year_Minimum_Errors.ipynb, updated other documents
Jul 2-6, 2019 · Version 1.2.2: split deeper_insight.docx into two documents, typeset them in LaTeX, uploaded each pdf produced from each TeX source, revised checking_calculations.xlsx, updated overview 
Jul 10-11, 2018 · Version 1.2.1: updated documents
Jun 21, 2018 · Version 1.2: redid deeper_insight.docx
Jun 13-15, 2018 · Version 1.1.1: updated Word documents, reorganized resources, added README
Mar 27, 2018 · Version 1.1: updated Word documents
Mar 21-22, 2018 · Version 1: initial uploads

gitlab-math.lua
-- GitLab uses a special format for specifying mathematics: $`[inline]`$ (example)
-- Normal format is, for example, $[inline]$
-- This pandoc filter will parse the GitLab format

function Math(el)
    if el.mathtype == "InlineMath" then
        if el.text:sub(1,1) == '`' and el.text:sub(#el.text) == '`' then
            local text = el.text:sub(2,#el.text-1)
            return pandoc.Math(el.mathtype, text)
        else
            local cont = pandoc.read(el.text)
            return { pandoc.Str("$") } .. cont.blocks[1].content .. { pandoc.Str("$") }
        end
    end
end

function CodeBlock(el)
    if el.classes[1] == "math" then
        return pandoc.Para({ pandoc.Math("DisplayMath", el.text) })
    end
end

-- Source: https://gist.github.com/lierdakil/00d8143465a488e0b854a3b4bf355cf6

SageMath/Ten-Year_Minimum_Errors copy.py
#
#  Ten-Year_Minimum_Errors.ipynb
#  Student Loans
#
#  Created by Ed Silkworth on 7/29/19.
#  Copyright © 2019-2020 Ed Silkworth. All rights reserved.
#

# for computing ten-year minimum payment errors
# open in SageMath Notebook

import multiprocessing as mp ; # for accelerating computations
import time # for timing computations
# for acceleration and timing, ``computations'' mean all processes involved in computing errors

# given
p_max = 10000 ; # default: 10000
p_min = 2000 ; # default: 2000
N = 40 ; # default: 40
# p will vary, see BO() and T() for more information
APR = 4.53 ; # default: 4.53
compound = 0 ; # 1 (yes) or 0 (no), default: 0
alpha = 1 # default: 1
# a = a_min, see BO() regarding a_min
# length of repayment in years and months, total payments, savings and change in savings are ignored

# set
delta_N = (p_max-p_min)/N ;

if compound == 0:
    i = APR/100/12
elif compound == 1: 
    i = pow( 1+APR/100/365.25 , 365.25/12 )-1
else: None
    
# test within $0.02 of a_min
test = 2 # see BO() regarding a_min, and see T() for running the test  

# check and round quantities to the nearest cent with a function
def CR(x): 
    
    if x*100-floor(x*100) > 0.499999 and x*100-floor(x*100) < 0.5:
        return round(x*100+1)/100
    else: 
        return round(x*100)/100

# compute the monthly principal balance and, consequently, monthly outstanding interest with a function
def BO(i, alpha, p, c): # see apply_async() and T() regarding p and c
    
    # set
    B = [p] ; # meaning B[0] = p
    O = [0] ; # meaning O[0] = 0
    m = 1
    
    if i > 0 and alpha == 0:
        a_min = ceil( B[0]/120 *100)/100
    elif i == 0:
        a_min = ceil( B[0]/120 *100)/100
    elif i > 0 and 0 < alpha <= 1:
        a_min = ceil( alpha*(B[0]*i)*pow(1+alpha*i,120) / (pow(1+alpha*i,120)-1) *100)/100 # case in which the numerical solution may deviate
    else: None
        
    # testing minimum
    a_min += 0.01*c
    
    while B[m-1] - (a_min - CR(alpha*B[m-1]*i)) > 0 and CR(a_min) != CR(alpha*B[0]*i): # do, except if B[1]=B[0]
        B.append(CR(B[m-1] - (a_min - CR(alpha*B[m-1]*i)))) ; # check whenever round, and check and round at end
        O.append(CR(O[m-1] + CR(B[m-1]*i) - CR(alpha*B[m-1]*i))) ;
        m += 1
    
    # if B[1]=B[0], remove potential error from consideration  
    if CR(a_min) == CR(alpha*B[0]*i):
        return # exit function and return nothing
    else: None
    
    # set
    n = m
    
    # set
    af = B[n-1] + CR(B[n-1]*i) + O[n-1] ;
    B.append(B[n-1] - (af - CR(B[n-1]*i) - O[n-1])) # = 0
    # O[n] = 0
    
    return B

# for running the test
def T(j, r, loop):
    
    # log potential errors in L
    if len( BO(i, alpha, p_min+delta_N*j, 0) )-1 <= 120: 
        L = [0/100] # meaning potentially no error exists
    else: L = []

    for c in r:
        if len( BO(i, alpha, p_min+delta_N*j, -(c+1)) )-1 <= 120:
            L.append(-(c+1)/100)
        else: None
        if len( BO(i, alpha, p_min+delta_N*j, c+1) )-1 <= 120:
            L.append((c+1)/100)            
        else: None    
        
    # compute any errors, and output them
    if float(min(L)) != 0:
        if float(min(L)) > 0:
            print "%s, +%.2f" %(p_min+delta_N*j, min(L))
        else: 
            print "%s, %.2f" %(p_min+delta_N*j, min(L))
    else: None
    
    # show progress periodically for large numbers of computations
    # without doing so and if errors are few, output may appear to stall
    if mod(p_min+delta_N*j,1000) == 0:
        print "\n(at p = %s)\n" %(p_min+delta_N*j)
    else: None
    
    # just for ones' information
    if loop == N+1:
        print "\n--\nOut of %s computation(s)" %(N+1)
    else: None
         
# for implementing multiprocessing
def apply_async():
    
    pool = mp.Pool()
    
    loop = 1 ;
    for j in range(N+1):
        pool.apply_async(T, args=(j, range(test), loop, )) ; # empty slot at end is required
        loop += 1

    pool.close() ;
    pool.join()

print "\nChecking for errors...\n(p, error)\n" # prints first

start = time.time()

if __name__ == "__main__":
    apply_async()                           
else: None

end = time.time()

print "\nCompleted in %.2fs" %(end-start)

Python/Ten-Year_Minimum_Errors.py
#
#  Ten-Year_Minimum_Errors.ipynb
#  Student Loans
#
#  Created by Ed Silkworth on 7/29/19.
#  Copyright © 2019-2020 Ed Silkworth. All rights reserved.
#

# for computing ten-year minimum payment errors

import multiprocessing as mp # for accelerating computations
import math # for certain mathematic functions
import time # for timing computations
# for acceleration and timing, ``computations'' mean all processes involved in computing errors

# given
p_max = 10000 # default: 10000
p_min = 2000 # default: 2000
N = 40 # default: 40
# p will vary, see BO() and T() for more information
APR = 4.53 # default: 4.53
compound = 0 # 1 (yes) or 0 (no), default: 0
alpha = 1 # default: 1
# a = a_min, see BO() regarding a_min
# length of repayment in years and months, total payments, savings and change in savings are ignored

def open():
    print("\nChecking for errors...\n(p, error)\n") # prints first

def close():
    end = time.time()
    print("\nCompleted in %.2fs" %(end-start))

# set
delta_N = (p_max-p_min)/N

if compound == 0:
    i = APR/100/12
elif compound == 1: 
    i = pow( 1+APR/100/365.25 , 365.25/12 )-1
else: None
    
# test within $0.02 of a_min
test = 2 # see BO() regarding a_min, and see T() for running the test  

# check and round quantities to the nearest cent with a function
def CR(x): 
    
    if x*100-math.floor(x*100) > 0.499999 and x*100-math.floor(x*100) < 0.5:
        return round(x*100+1)/100
    else: 
        return round(x*100)/100

# compute the monthly principal balance and, consequently, monthly outstanding interest with a function
def BO(i, alpha, p, c): # see apply_async() and T() regarding p and c
    
    # set
    B = [p] # meaning B[0] = p
    O = [0] # meaning O[0] = 0
    m = 1
    
    if i > 0 and alpha == 0:
        a_min = math.ceil( B[0]/120 *100)/100
    elif i == 0:
        a_min = math.ceil( B[0]/120 *100)/100
    elif i > 0 and 0 < alpha <= 1:
        a_min = math.ceil( alpha*(B[0]*i)*pow(1+alpha*i,120) / (pow(1+alpha*i,120)-1) *100)/100 # case in which the numerical solution may deviate
    else: None
        
    # testing minimum
    a_min += 0.01*c
    
    while B[m-1] - (a_min - CR(alpha*B[m-1]*i)) > 0 and CR(a_min) != CR(alpha*B[0]*i): # do, except if B[1]=B[0]
        B.append(CR(B[m-1] - (a_min - CR(alpha*B[m-1]*i)))) # check whenever round, and check and round at end
        O.append(CR(O[m-1] + CR(B[m-1]*i) - CR(alpha*B[m-1]*i)))
        m += 1
    
    # if B[1]=B[0], remove potential error from consideration  
    if CR(a_min) == CR(alpha*B[0]*i):
        return # exit function and return nothing
    else: None
    
    # set
    n = m
    
    # set
    af = B[n-1] + CR(B[n-1]*i) + O[n-1]
    B.append(B[n-1] - (af - CR(B[n-1]*i) - O[n-1])) # = 0
    # O[n] = 0
    
    return B

# for running the test
def T(j, r, loop):
    
    # just for ones' information
    if loop == 1:
        open()

    # log potential errors in L
    if len( BO(i, alpha, p_min+delta_N*j, 0) )-1 <= 120: 
        L = [0/100] # meaning potentially no error exists
    else: L = []

    for c in r:
        if len( BO(i, alpha, p_min+delta_N*j, -(c+1)) )-1 <= 120:
            L.append(-(c+1)/100)
        else: None
        if len( BO(i, alpha, p_min+delta_N*j, c+1) )-1 <= 120:
            L.append((c+1)/100)            
        else: None    
        
    # compute any errors, and output them
    if float(min(L)) != 0:
        if float(min(L)) > 0:
            print("%s, +%.2f" %(p_min+delta_N*j, min(L)))
        else: 
            print("%s, %.2f" %(p_min+delta_N*j, min(L)))
    else: None
    
    # show progress periodically for large numbers of computations
    # without doing so and if errors are few, output may appear to stall

    factor = math.floor((p_min+delta_N*j)/1000)
    if (p_min+delta_N*j) >= 1000*factor and (p_min+delta_N*(j-1)) < 1000*factor:
        print("\n(at p = %s)\n" %(1000*factor))
    else: None
    
    # just for ones' information
    if loop == N+1:
        print("\n--\nOut of %s computation(s)" %(N+1))
        close()
    else: None
         
# for implementing multiprocessing
def apply_async():
    
    pool = mp.Pool()
    
    loop = 1
    for j in range(N+1):
        pool.apply_async(T, args=(j, range(test), loop, )) # empty slot at end is required
        loop += 1

    pool.close()
    pool.join()

start = time.time()

if __name__ == "__main__":
    apply_async()                           
else: None

# end = time.time() placed in close()